light-node/src/commands/start.rs

```diff
@@ -0,0 +1,161 @@
+
+        subjective_init(config.subjective_init.height, vals_hash, &mut store, &req).unwrap();
+
+        loop {
+            let latest_sh = (&req).signed_header(0).unwrap();
+            let latest_peer_height = latest_sh.header().height();
+
+            let latest_trusted = store.get(0).unwrap();
+            let latest_trusted_height = latest_trusted.last_header().header().height();
+
+            // only bisect to higher heights
+            if latest_peer_height <= latest_trusted_height {
+                std::thread::sleep(Duration::new(1, 0));
+                continue;
+            }
+
+            println!(
+                "attempting bisection from height {:?} to height {:?}",
+                latest_trusted_height, latest_peer_height,
+            );
+
+            let now = SystemTime::now();
+            lite::verify_bisection(
+                latest_trusted.to_owned(),
+                latest_peer_height,
+                TrustThresholdFraction::default(), // TODO
+                config.trusting_period,
+                now,
+                &req,
+            )
+            .unwrap();
+
+            println!("Succeeded bisecting!");
+
+            // notifications ?
+
+            // sleep for a few secs ?
+        }
+    }
+}
+
+impl config::Override<LightNodeConfig> for StartCmd {
+    // Process the given command line options, overriding settings from
+    // a configuration file using explicit flags taken from command-line
+    // arguments.
+    fn override_config(
+        &self,
+        mut config: LightNodeConfig,
+    ) -> Result<LightNodeConfig, FrameworkError> {
+        if !self.rpc_addr.is_empty() {
+            config.rpc_address = self.rpc_addr.to_owned();
+        }
+
+        Ok(config)
+    }
+}
+
+/*
+ * The following is initialization logic that should have a
+ * function in the lite crate like:
+ * `subjective_init(height, vals_hash, store, requester) -> Result<(), Error`
+ * it would fetch the initial header/vals from the requester and populate a
+ * trusted state and store it in the store ...
+ * TODO: this should take traits ... but how to deal with the State ?
+ * TODO: better name ?
+*/
+fn subjective_init(
+    height: Height,
+    vals_hash: Hash,
+    store: &mut MemStore,
+    req: &RPCRequester,
+) -> Result<(), Error> {
+    if store.get(height).is_ok() {
+        // we already have this !
+        return Ok(());
+    }
+
+    // check that the val hash matches
+    let vals = req.validator_set(height)?;
+
+    if vals.hash() != vals_hash {
+        // TODO
+        panic!("vals hash dont match")
+    }
+
+    let signed_header = req.signed_header(height)?;
+
+    // TODO: validate signed_header.commit() with the vals ...
+
+    let next_vals = req.validator_set(height + 1)?;
+
+    // TODO: check next_vals ...
+
+    let trusted_state = &State::new(&signed_header, &next_vals);
+
+    store.add(trusted_state.to_owned())?;
+
+    Ok(())
+}
+
+fn block_on<F: Future>(future: F) -> F::Output {
```

Comment:
Have you considered not building a new runtime on each call?

You could probably do something like:

```rust
lazy_static! {
    pub static ref RUNTIME: Runtime = { Builder::new()
        .basic_scheduler()
        .enable_all()
        .build()
        .unwrap() };
}

fn block_on<F: Future>(future: F) -> F::Output {
    RUNTIME.block_on(future)
}
```

I can't tell if this is called from multiple threads or not, if so, you could just put a `Mutex` around the `Runtime`.

Reply:
You might want to check out [`abscissa_tokio`](https://github.com/iqlusioninc/abscissa/tree/develop/tokio) which allows you to use the Tokio runtime as a component (so you can spawn things onto it during application startup if you so desire) and boot the runtime using:

https://docs.rs/abscissa_tokio/0.5.1/abscissa_tokio/#add-tokiocomponent-to-your-abscissa-application

```rust
use crate::application::APPLICATION;

impl Runnable for StartCmd {
   fn run(&self) {
       abscissa_tokio::run(&APPLICATION, async {
           println!("now running inside the Tokio runtime");
       });
   }
}
```

Reply:
It seems the current approach is to block on the future in the current thread, I assume the solution linked to above spawns the future onto Tokio asynchronously instead? It might be useful and it would seem like a change from the current approach.

Reply:
I think you are right @gterzian. 

Note that currently there is only one call to this particular `block_on` method and runtime. But yes, we need to do this properly throughout the repo. We currently [copy](https://github.com/interchainio/tendermint-rs/pull/116#issue-357829136)&pasted [this](https://github.com/interchainio/tendermint-rs/pull/85#issuecomment-564437616) throughout the repo when async/await was introduced [here](https://github.com/interchainio/tendermint-rs/pull/85).

Currently the [rpc integration tests](https://github.com/interchainio/tendermint-rs/blob/6df2a927623f4c50869249c89a4502f750f246f5/tendermint/tests/integration.rs#L16-L23), above start command and the [requester](https://github.com/interchainio/tendermint-rs/blob/9c4fd2fb953aab71d4073c4fac106027073b3b5b/light-node/src/requester.rs#L51-L58) use this approach. Maybe worth to have a closer look into this in a separate issue, too?

Reply:
@gterzian the semantics are the same, but yes it would use the Tokio executor.

You can make it synchronous by replacing `block_on` with `await` to wait for the future to complete, otherwise the semantics are the same (except with async/await, you can add async behavior in the future)

Reply:
Yes I think it's worth discussing the overall approach in a separate issue(it shouldn't matter much to be creating a new runtime on each call for now). 

Yes if we call `await` from inside an async context we'll get the equivalent of the current `block_on`, and I think the tricky part is `block_on` is not called from an async context? 

So if we run the entire application on Tokio, we could indeed use `await` instead of `block_on`. 

Probably best to keep the current `block_on`, and then figure out how we want to run the light client.

Reply:
https://github.com/interchainio/tendermint-rs/issues/167